### ISSUE TITLE
Some meshes and UVs are badly distorted

### ISSUE THREAD

**Issue by John-Nagle** [BVE-Reborn/rend3#239]
After applying the workaround for Issue #237, the basic test scene renders with Rend3 0.5.0, with problems.

1. A few meshes are wrong, with vertices way out of place.
2. Some meshes have wrong UVs.
3. During mesh loading, frame rate is very slow. Some frames take several seconds to render. Once all meshes have loaded, texture loading is as fast as before.

None of this happened with the previous Rend3. Changes on my side were confined to those required by the PbrMaterial and Camera transform changes.

![rend0 5 0badmesh1](https://user-images.githubusercontent.com/942608/137534271-8e68cf86-fafa-4f1d-afa5-8aab93615780.png)
**Bad mesh.**

![rend0 5 0baduvs1](https://user-images.githubusercontent.com/942608/137534289-5b8f2195-9c8a-4789-a24d-678611e1393a.png)
**Bad UVs.**

**Comment by John-Nagle** [comment]
Performance while meshes are being loaded is very, very slow.
```
00004 frames over 01.12s. Min: 139.59ms; Average: 280.72ms; 95%: 413.14ms; 99%: 413.14ms; Max: 413.14ms; StdDev: 105.09ms
00053 frames over 01.01s. Min: 08.81ms; Average: 19.04ms; 95%: 36.04ms; 99%: 36.73ms; Max: 36.73ms; StdDev: 06.76ms
00060 frames over 01.00s. Min: 14.10ms; Average: 16.69ms; 95%: 18.51ms; 99%: 18.91ms; Max: 18.91ms; StdDev: 01.07ms
00061 frames over 01.02s. Min: 14.12ms; Average: 16.66ms; 95%: 18.74ms; 99%: 19.51ms; Max: 19.51ms; StdDev: 01.19ms
00060 frames over 01.00s. Min: 14.99ms; Average: 16.71ms; 95%: 18.53ms; 99%: 18.63ms; Max: 18.63ms; StdDev: 01.10ms
00061 frames over 01.01s. Min: 12.88ms; Average: 16.64ms; 95%: 18.76ms; 99%: 20.46ms; Max: 20.46ms; StdDev: 01.37ms
00060 frames over 10.61s. Min: 12.10ms; Average: 176.91ms; 95%: 19.51ms; 99%: 9621.73ms; Max: 9621.73ms; StdDev: 1229.61ms
00002 frames over 17.97s. Min: 10.92ms; Average: 8989.66ms; 95%: 17968.40ms; 99%: 17968.40ms; Max: 17968.40ms; StdDev: 8978.74ms
00002 frames over 37.60s. Min: 16.29ms; Average: 18815.40ms; 95%: 37614.52ms; 99%: 37614.52ms; Max: 37614.52ms; StdDev: 18799.12ms
00003 frames over 17.83s. Min: 08.55ms; Average: 5947.38ms; 95%: 17817.40ms; 99%: 17817.40ms; Max: 17817.40ms; StdDev: 8393.37ms
```

**Comment by John-Nagle** [comment]
WGPU error and crash after about an hour of very slow mesh loading.

```
23:07:36 [ERROR] Handling wgpu errors as fatal by default
thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_bind_group
      note: label = `D2 texture bg count 8194`
    binding count declared with at most 8192 items, but 8194 items were provided

', /home/john/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.11.0/src/backend/direct.rs:2195:5
stack backtrace:
Asset queue, 4270 entries.
   0: rust_begin_unwind
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:515:5
   1: std::panicking::begin_panic_fmt
             at /rustc/c8dfcfe046a7680554bf4eb612bad840e7631c4b/library/std/src/panicking.rs:457:5
   2: wgpu::backend::direct::default_error_handler
   3: core::ops::function::Fn::call
   4: <wgpu::backend::direct::Context as wgpu::Context>::device_create_bind_group
   5: wgpu::Device::create_bind_group
   6: rend3::resources::texture::create_bind_group
   7: rend3::resources::texture::TextureManager::ready
   8: rend3::renderer::render::render_loop
   9: slscenetester::displaystate::{{closure}}
  10: winit::platform_impl::platform::x11::EventLoop<T>::run
  11: winit::platform_impl::platform::EventLoop<T>::run
  12: winit::event_loop::EventLoop<T>::run
  13: slscenetester::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Translucency cutout of RenderTextureIdent { uuid_key_range: UuidKeyRange { uuid_key: (5, 7972fdc4-5436-c046-09f2-8acc26d5508a), highest_lod_available: 1 }, texture_type: DiffuseTexture }: 99
Translucency cutout of RenderTextureIdent { uuid_key_range: UuidKeyRange { uuid_key: (5, 7972fdc4-5436-c046-09f2-8acc26d5508a), highest_lod_available: 1 }, texture_type: DiffuseTexture }: 35
Translucency cutout of RenderTextureIdent { uuid_key_range: UuidKeyRange { uuid_key: (5, 7972fdc4-5436-c046-09f2-8acc26d5508a), highest_lod_available: 1 }, texture_type: DiffuseTexture }: 34
Asset queue, 4260 entries.
Segmentation fault (core dumped)
```

**Comment by John-Nagle** [comment]
More info: "Bad mesh" problem above only appears when meshes are dropped before being used. So, this is similar to the problem in #237 with dropping materials before using them. 

This does NOT fix the bad UV problem or the slow loading problem.

**Comment by John-Nagle** [comment]
More info: slow loading problem was due to logging far too much debug info. With debug level at "warn", refresh does not drop below 60 FPS during content loading. This is better than Rend3 0.1.x, which slowed down during content loading from another thread.

Still have some bad UV problems. This may be related to the changes involving the new uv_texture0 and uv_texture1. Unclear on how to use those features.  Currently putting the  transform previously put into "texture" into both the "uv_texture0" and "uv_texture1" fields, since that's what the GLTF loader does. Also tried just leaving the identity transform in "uv_texture1". No difference between those two approaches. 

Worth noting is that some transforms I use may contain negative scale factors. This worked fine in 0.1.x and is needed, because the content I get from Second Life uses that feature. May or may not be related.

**Comment by cwfitzgerald** [comment]
Useful info!

> Also tried just leaving the identity transform in "uv_texture1".

This is what you should do for now. This was added to support baking/reading from a lightmap, but the rest of the material system hasn't caught up to actually let you do this, so both the second set of UVs, and their transform is entirely unused in practice.

> Worth noting is that some transforms I use may contain negative scale factors. 

This should still work fine, I always use wrapping addressing on textures, so even if they have some truely funky transforms, it will work okay.

I have a hunch this is also related to #237 just manifesting differently.

**Comment by John-Nagle** [comment]
More info: The bad UV problem seems to be transform-related, not UV-related. I'm seeing some objects with the proper texture rotated 90 or 180 degrees. I am not seeing garbage UVs, just rotated or mis-scaled textures. 

Sometimes a group of identical objects has different texture rotations for objects in the group. It's as if one object has the transform for some other object.

I'm starting to suspect that most of these problems come from the same underlying index-related bug.

**Comment by John-Nagle** [comment]
More info:

Here's the bad texture rotation problem. Note that some signs are rotated 90° or 180°. UVs are OK, but texture rotation matrices are not. 

I may be doing something wrong with the new two texture transform matrices. This worked fine with the old one-matrix system.

![badtexturerotation](https://user-images.githubusercontent.com/942608/138139493-31a07181-4096-4cc0-8b15-dc9f92923944.png)

**Comment by John-Nagle** [comment]
As of Rend3 0.2.1, the only remaining problem in this issue is the "bad texture rotation problem". Sometimes the scale is off. It's as if the texture rotation matrix for the wrong object is sometimes used. 
![texture-transform-bug-1](https://user-images.githubusercontent.com/942608/138583441-0ebe59af-7e46-4a2a-bc32-4096ad4f42aa.png)
**Note map and diploma.**

**Comment by cwfitzgerald** [comment]
I think my normal test scene might be experiencing this, I was just unaware due to its simplicity. Investigating...

Additionally I fixed the flicker when changing textures, that one reproed easily.

**Comment by cwfitzgerald** [comment]
Closed via #245 and released in rend3(-pbr) 0.2.2
